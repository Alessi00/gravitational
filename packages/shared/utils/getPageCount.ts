import type { AgentResponse, AgentKind } from 'teleport/services/agents';

export default function getPageCount({ page, fetchedData, limit = 15 }: Props) {
  // Calculate counts for our resource list.
  let from = 0;
  let to = 0;
  let total = 0;
  if (fetchedData.totalCount) {
    from = page.index * limit + 1;
    to = from + fetchedData.agents.length - 1;
    total = fetchedData.totalCount;
  }
  return {
    from,
    to,
    total,
  };
}

// Page keeps track of our current agent list
//  start keys and current position.
export type Page = {
  // keys are the list of start keys collected from
  // each page fetch.
  keys: string[];
  // index refers to the current index the page
  // is at in the list of keys.
  index: number;
};

type Props = {
  page: Page;
  fetchedData: AgentResponse<AgentKind>;
  limit?: number;
};
