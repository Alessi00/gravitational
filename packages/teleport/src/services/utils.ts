/**
 * Copyright 2021 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Label, TagDict } from './types';

// makeResourceTags takes a list of resource labels and converts them to
// tag format `name: value`. Unique tags will be stored in the tagMap.
export function makeResourceTags(labels: Label[] = [], tagDict: {}) {
  return labels.map(label => {
    const tag = `${label.name}: ${label.value}`;
    if (!tagDict[tag]) {
      tagDict[tag] = label;
    }

    return tag;
  });
}
