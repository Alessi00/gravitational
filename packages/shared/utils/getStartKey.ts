import { Page } from './getPageCount';

export default function getStartKey({ dir = '', page }: Props) {
  switch (dir) {
    case 'prev':
      return page.keys[page.index - 1];
    case 'next':
      return page.keys[page.index + 1];
    default:
      return undefined;
  }
}

type Props = {
  dir: '' | 'prev' | 'next';
  page: Page;
};
