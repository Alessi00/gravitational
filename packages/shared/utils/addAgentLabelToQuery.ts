import { AgentFilter, AgentLabel } from 'teleport/services/agents';

export function addAgentLabelToQuery(filter: AgentFilter, label: AgentLabel) {
  const queryParts = [];

  // Add existing query
  if (filter.query) {
    queryParts.push(filter.query);
  }

  // If there is an existing simple search,
  // convert it to predicate language and add it
  if (filter.search) {
    queryParts.push(`search("${filter.search}")`);
  }

  // Create the label query.
  queryParts.push(`labels["${label.name}"] == "${label.value}"`);

  return queryParts.join(' && ');
}
