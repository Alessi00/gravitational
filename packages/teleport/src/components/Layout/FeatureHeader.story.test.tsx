/**
 * Copyright 2022 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  NoMessages,
  OneMessage,
  MultipleMessages,
} from './FeatureHeader.story';
import { render, waitFor } from 'design/utils/testing';

test('rendering of header without messages', async () => {
  const { container } = render(<NoMessages />);

  await waitFor(() => document.querySelector('div'));
  expect(container).toMatchSnapshot();
});

test('rendering of header without one message', async () => {
  const { container } = render(<OneMessage />);

  await waitFor(() => document.querySelector('div'));
  expect(container).toMatchSnapshot();
});

test('rendering of header with multiple messages', async () => {
  const { container } = render(<MultipleMessages />);

  await waitFor(() => document.querySelector('div'));
  expect(container).toMatchSnapshot();
});
