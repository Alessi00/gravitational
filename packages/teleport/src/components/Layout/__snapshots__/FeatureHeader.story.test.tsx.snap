// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering of header with multiple messages 1`] = `
.c0 {
  box-sizing: border-box;
  margin-bottom: 24px;
}

.c5 {
  box-sizing: border-box;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.c6 {
  box-sizing: border-box;
  display: flex;
  align-items: center;
}

.c7 {
  display: inline-block;
  transition: color .3s;
  margin-right: 8px;
  color: #FFFFFF;
  font-size: 16px;
}

.c3 {
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 2px;
  box-sizing: border-box;
  box-shadow: 0 1px 4px rgba(0,0,0,0.24);
  margin: 0 0 24px 0;
  min-height: 40px;
  padding: 8px 16px;
  overflow: auto;
  word-break: break-word;
  line-height: 1.5;
  margin-bottom: 0;
  background: #f50057;
  color: #FFFFFF;
}

.c3 a {
  color: #FFFFFF;
}

.c8 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 600;
  margin: 0px;
}

.c9 {
  color: #03a9f4;
  font-weight: normal;
  background: none;
  text-decoration: underline;
  text-transform: none;
}

.c1 {
  box-sizing: border-box;
  display: flex;
  align-items: center;
  flex-shrink: 0;
  border-bottom: 1px solid #1C254D;
  height: 56px;
  margin-left: -40px;
  margin-right: -40px;
  padding-left: 40px;
  padding-right: 40px;
  position: relative;
}

.c4 {
  box-shadow: 0 1px 4px rgb(0 0 0 / 50%);
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  margin: 0px;
  white-space: nowrap;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Applications
      </div>
    </div>
    <div
      class="c3 c4"
      display="flex"
      kind="danger"
    >
      <div
        class="c5"
        width="100%"
      >
        <div
          class="c6"
        >
          <span
            class="c7 icon icon-info_outline "
            color="light"
            font-size="3"
          />
          <div
            class="c8"
          >
            error
          </div>
        </div>
        <a
          class="c9"
          rel="noreferrer"
          style="font-size: 12px; cursor: pointer; min-width: 44px;"
        >
          Dismiss
        </a>
      </div>
    </div>
    <div
      class="c3 c4"
      display="flex"
      kind="danger"
    >
      <div
        class="c5"
        width="100%"
      >
        <div
          class="c6"
        >
          <span
            class="c7 icon icon-info_outline "
            color="light"
            font-size="3"
          />
          <div
            class="c8"
          >
            second error
          </div>
        </div>
        <a
          class="c9"
          rel="noreferrer"
          style="font-size: 12px; cursor: pointer; min-width: 44px;"
        >
          Dismiss
        </a>
      </div>
    </div>
    <div
      class="c3 c4"
      display="flex"
      kind="danger"
      style="border-radius: 0 0 6px 6px;"
    >
      <div
        class="c5"
        width="100%"
      >
        <div
          class="c6"
        >
          <span
            class="c7 icon icon-info_outline "
            color="light"
            font-size="3"
          />
          <div
            class="c8"
          >
            another error
          </div>
        </div>
        <a
          class="c9"
          rel="noreferrer"
          style="font-size: 12px; cursor: pointer; min-width: 44px;"
        >
          Dismiss
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`rendering of header without messages 1`] = `
.c0 {
  box-sizing: border-box;
  margin-bottom: 24px;
}

.c1 {
  box-sizing: border-box;
  display: flex;
  align-items: center;
  flex-shrink: 0;
  border-bottom: 1px solid #1C254D;
  height: 56px;
  margin-left: -40px;
  margin-right: -40px;
  padding-left: 40px;
  padding-right: 40px;
  position: relative;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  margin: 0px;
  white-space: nowrap;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Applications
      </div>
    </div>
  </div>
</div>
`;

exports[`rendering of header without one message 1`] = `
.c0 {
  box-sizing: border-box;
  margin-bottom: 24px;
}

.c5 {
  box-sizing: border-box;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.c6 {
  box-sizing: border-box;
  display: flex;
  align-items: center;
}

.c7 {
  display: inline-block;
  transition: color .3s;
  margin-right: 8px;
  color: #FFFFFF;
  font-size: 16px;
}

.c3 {
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 2px;
  box-sizing: border-box;
  box-shadow: 0 1px 4px rgba(0,0,0,0.24);
  margin: 0 0 24px 0;
  min-height: 40px;
  padding: 8px 16px;
  overflow: auto;
  word-break: break-word;
  line-height: 1.5;
  margin-bottom: 0;
  background: #f50057;
  color: #FFFFFF;
}

.c3 a {
  color: #FFFFFF;
}

.c8 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 600;
  margin: 0px;
}

.c9 {
  color: #03a9f4;
  font-weight: normal;
  background: none;
  text-decoration: underline;
  text-transform: none;
}

.c1 {
  box-sizing: border-box;
  display: flex;
  align-items: center;
  flex-shrink: 0;
  border-bottom: 1px solid #1C254D;
  height: 56px;
  margin-left: -40px;
  margin-right: -40px;
  padding-left: 40px;
  padding-right: 40px;
  position: relative;
}

.c4 {
  box-shadow: 0 1px 4px rgb(0 0 0 / 50%);
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  margin: 0px;
  white-space: nowrap;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Applications
      </div>
    </div>
    <div
      class="c3 c4"
      display="flex"
      kind="danger"
      style="border-radius: 0 0 6px 6px;"
    >
      <div
        class="c5"
        width="100%"
      >
        <div
          class="c6"
        >
          <span
            class="c7 icon icon-info_outline "
            color="light"
            font-size="3"
          />
          <div
            class="c8"
          >
            error
          </div>
        </div>
        <a
          class="c9"
          rel="noreferrer"
          style="font-size: 12px; cursor: pointer; min-width: 44px;"
        >
          Dismiss
        </a>
      </div>
    </div>
  </div>
</div>
`;
